### ISSUE TITLE
Splitting gui vs non-gui tests

### ISSUE THREAD

**Issue by sofroniewn** [napari/napari#1363]
## 🚀 Feature
Make it possible clear how to run gui/ non-gui tests independently of each other

## Motivation

Right now you can run all the tests together which can take multiple minutes and which sometimes causes Widgets to pop up on the screen, which can give the impression your machine has been "taken over", and which can delay development.

## Pitch

It should be possible to run all our non gui tests together

## Additional context

Note that one piece of complexity here is that not all Qt tests actually launch the viewer. I think there are some ways to run Qt tests which prevent the viewer from launching that @tlambert03 implemented, but that the viewer needs to launch for some of them.

One option would be just to have all the qt tests and non-qt tests grouped.

Right now the qt tests are found in
- https://github.com/napari/napari/tree/master/napari/_tests
- https://github.com/napari/napari/tree/master/napari/_qt/_tests
- https://github.com/napari/napari/tree/master/napari/_vispy

Non-qt tests are found in
- https://github.com/napari/napari/tree/master/napari/components/_tests
- https://github.com/napari/napari/tree/master/napari/layers/_tests (though it looks like the test_dask_layers.py might violate this and need to be moved to `napari/_tests`)
- https://github.com/napari/napari/tree/master/napari/utils/_tests
- https://github.com/napari/napari/tree/master/napari/plugins/_tests

If any qt tests have leaked into those locations they should probably be moved.

One option would be to use markers in pytest https://docs.pytest.org/en/latest/example/markers.html which can then be included/ excluded (h/t @pwinston) Another is to see if there are ways to make aliases to just run tests in the non-qt folders and make sure our organization is obeyed.

Thoughts @jni @tlambert03 @ziyangczi

**Comment by tlambert03** [comment]
As I mentioned when @pwinston brought this up on zulip, I _think_ (but am not sure) that a lot of our "with gui" tests could probably be refactored to not require showing the gui.  Many of the tests that use `viewer_factory(show=True)` are only that way because when I originally made most of the tests not use the gui (with the `show` option to `viewer_factory` in #1014), I had to use `show=True` for a handful of them to pass.  But I remember thinking at the time that they didn't really _need_ to show a viewer, it was just the easiest path at the time, and not the point of that PR.

Some will definitely need to be shown (like screenshots, and those that click on specific parts of the screen) ... but before we focus on splitting the ones that require the gui into a different test, it might be better to put time into minimizing the number of tests that actually need the gui.

**Comment by pwinston** [comment]
If we can reduce the number of "gui" tests while still testing all the same stuff that seems great. And it would lessen the number of tests that need to be marked "GUI". 

The pytest [markers](https://docs.pytest.org/en/latest/example/markers.html) feature seems like it was design for something like that. Marking tests of a certain type, and being able to run marked tests or un-marked tests.

Both of these goals marking GUI tests and reducing the number of GUI tests feel like pretty low priority. But they seem worth doing at some point.

**Comment by liu-ziyang** [comment]
I looked into this a bit in depth, and have 2 potential solutions:
1. less disruptive: rename a few integration test to have "gui" postfix under napari/_test, and then developer can then run  `pytest -k "not qt and not vispy and not gui` to exclude these test
2. more disruptive: add "gui" mark to all tests that have gui components, and then developer can run test with `pytest -m "not gui" or pytest -m gui`

both have its advantage and disadvantage in how the name/tag is used to group tests, and have implications of how we are planning to organize tests in the future. That being said, I still hope running only the relevant test during the dev iteration cycle, instead of always run the whole test suit would be more scalable in any case

**Comment by sofroniewn** [comment]
> 1. less disruptive: rename a few integration test to have "gui" postfix under napari/_test, and then developer can then run pytest -k "not qt and not vispy and not gui to exclude these test
> 2. more disruptive: add "gui" mark to all tests that have gui components, and then developer can run test with pytest -m "not gui" or pytest -m gui

@jni @tlambert03 do you have a preference for 1. or 2.? 

I'm pretty open to either. Option 1. seems like a reasonable place to start. Here are the relevant docs on the `-k` option https://docs.pytest.org/en/stable/example/markers.html#using-k-expr-to-select-tests-based-on-their-name in case anyone wants to read more. We can always still move to option 2. later and it will be clear then the we should add the gui mark to tests which have gui in the name (or qt or vispy!).

Thanks for presenting these choices @ziyangczi!!

**Comment by jni** [comment]
Same view as you @sofroniewn

**Comment by tlambert03** [comment]
i don't really love either approach.  I think it avoids addressing the actual issue, which is the test itself.  I just pushed #1372 for your consideration.

**Comment by liu-ziyang** [comment]
Love the PR! though I think we are addressing two different root issue really, though both would work for this particular issue. What I was leaning towards was to have a more effective grouping of tests, be it gui, database connection, or w/e, and the end goal is to make unit test more scalable for developers to iterate. Imaging we are changing code related to gui only, then we can only run all unit tests related to that, instead of a whole test suite. Right now it is probably less an issue where unit tests takes ~3min, but when the test grows I think it would not be efficient to iterate without groupings

**Comment by tlambert03** [comment]
I guess my feeling on that is we _already_ should have that by nature of our code structure.  Unit tests are (or at least should be) in `_tests` subdirectories of the things they are testing.  We don't have any database resources or anything like that... but if we eventually do, I would hope that we'd put additions with new infrastructure dependencies in isolated subdirectories as well.

nobody has to run the whole test suite:  I almost always run `pytest napari/_qt` or `pytest napari/components`, etc... (unless I'm specifically wanting to know that I didn't break anything anywhere).  or, [a good tip for quickly testing with coverage locally](https://github.com/napari/napari/pull/871#issuecomment-573359637)...  If people are running the full test suite all the time, maybe we need to add some testing tips to our contributing docs?

In other words, before we start adding internal conventions on _how_ we test things, like "1) name your tests xyz, or mark it with xyz 2) run your test with `pytest -k "xyz and not abc"`", I would say let's maintain the code itself in a better state and use already-existing pytest strategies for running subsets of our tests.  I agree with @sofroniewn when he said: "If any qt tests have leaked into those locations they should probably be moved.

**Comment by sofroniewn** [comment]
> I would hope that we'd put additions with new infrastructure dependencies in isolated subdirectories as well.

Agreed, and we should be clear about what is where.

> I agree with @sofroniewn when he said: "If any qt tests have leaked into those locations they should probably be moved.

I can go through and do this myself tomorrow.

> maybe we need to add some testing tips to our contributing docs?

yes that probably is good too.

I think the combination of #1372, me going through our tests tomorrow to move any that are in the wrong place, and then @tlambert03 and I adding some more helpful instructions to our contributing guide about how tests are organized and can be run selectively would be good.

I think at this point we might just want to pull out a whole new document just for our tests and link to that from the contributing guide.

Then we can close this issue and consider it done. We can wait on further renaming @ziyangczi. I will ping people when I have that PR ready and then ask people to review and check the organization (right now I think it's only one file that needs to move).
